### PR TITLE
Refactor code: Create a class for the Image-Manager

### PR DESCRIPTION
**Pull request by sihensel**
Closes https://github.com/osism/openstack-image-manager/issues/228 and https://github.com/osism/openstack-image-manager/issues/213.

Create a single class that contains the whole functionality of the
Image-Manager which allows for easy exporting.
The Image-Manager can be invoked by running the `main()` method.
All unittests are updated accordingly.

The requirements for `python-glanceclient` and `os-client-config` are
removed, as it was forgotten in https://github.com/osism/openstack-image-manager/pull/236.

Do not delete image properties that start with `os_*` since they are set by glance.

Signed-off-by: Simon Hensel <simon.hensel@inovex.de>

### CODE REVIEW

**Comment by berendt** [comment]
@ra-rau Please test this one.

**Review by ra-rau — APPROVED** [review]
The changes work at my test-instance.

**Comment by horazont** [comment]
(Context: This came up in SCS team 1 today and I was invited to comment here, so I hope this is helpful.)

TL;DR: If the goal is reusability, then having individual functions with clear arguments (as they were before) seems more sensible to me.

That way, the image manager becomes also a library which can be used, instead of a monolithic class with non-pure dependencies (such as the CONF) and passing state between methods implicitly (via attributes).

Many of the functions have been transformed from taking a "context" argument (the connection) plus the arguments describing the operation (image name etc.) into taking self + the arguments taking the operation. That is an indicator that they should be free functions which should just take the single context argument.

In other cases, some of the function arguments have been moved to object attributes, even though they are part of transient state, making the object a kind of "register" to store state (`sorted_versions` is one such thing). `sorted_versions` is only used by `process_images` and `rename_images`, which are directly called from `main()`. If `sorted_versions` was instead passed as argument to `process_images`, it could in fact be used by someone instantiating the class; otherwise, this attribute needs to be set beforehand which is weird.

If the functions are transformed back to only use arguments or inherent object state, they could easily be transformed back to functions (based on the argument above), removing the mental overhead of using a class (and its potentially non-obvious state).

Having purer functions (no global/shared state beyond what is passed as arguments and what is returned) has the following advantages:

- Easier testing: no need to mock global state, passing mock objects as arguments is sufficient
- Easier reuse: no need to ensure that some shared state is correct; the arguments of the functions are sufficient to make them work correctly.

**Comment by sihensel** [comment]
@horazont Thanks for the comment, that's a good insight. I understand your point of view and agree with it. Pure, self-contained functions make code reuse and testing (slightly) easier and more straightforward.

The main advantage of a class is the reusability of class variables, which reduces the number or arguments a function has to take, since they can make use of a shared state (e.g. by moving the connection to `self`). But then, the connection should be set in the constructor instead of `main()`, so that it is available as soon as the object gets created.

In the case of `sorted_versions`, the usage is indeed awkward and could be improved.
If all function parameters are made consistent by using self + all arguments prescribing the operation, all functions work independent of each other (just like when using pure functions), since the shared state containing `CONF` and `conn` is established when the object is created.

In the end it comes down to which design principle is preferred, the code works just as well without using a class.
